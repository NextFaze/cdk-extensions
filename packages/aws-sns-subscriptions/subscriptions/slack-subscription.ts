import { Runtime } from '@aws-cdk/aws-lambda';
import { NodejsFunction } from '@aws-cdk/aws-lambda-nodejs';
import {
  ITopic,
  ITopicSubscription,
  SubscriptionProtocol,
  TopicSubscriptionConfig,
  SubscriptionFilter,
} from '@aws-cdk/aws-sns';
import { StringParameter } from '@aws-cdk/aws-ssm';
import { Construct, Duration } from '@aws-cdk/core';
import path from 'path';
// import { Names } from '@aws-cdk/core';

export interface ISlackConfigParam {
  channelId: string;
  channelName: string;
  authToken: string;
}

export interface ISlackSubscriptionProps {
  channelName: string;
  filterPolicy?: { [key: string]: SubscriptionFilter };
}

export class SlackSubscription implements ITopicSubscription {
  constructor(private props: ISlackSubscriptionProps) {}

  bind(topic: ITopic): TopicSubscriptionConfig {
    const scope = this._resolveScopeFromTopic(topic);

    const configParam = new StringParameter(scope, 'ConfigParameter', {
      stringValue: JSON.stringify({
        channelId: '<channel-id>',
        channelName: this.props.channelName,
        authToken: '<SECURE_TOKEN>',
      } as ISlackConfigParam),
      description: 'Slack configuration parameter',
    });

    const slackHandler = new NodejsFunction(scope, 'Handler', {
      runtime: Runtime.NODEJS_12_X,
      // slightly increase duration in case we need to \
      // scan all slack channels to find the one we are looking for
      timeout: Duration.minutes(3),
      entry: path.resolve(
        __dirname,
        '../handlers/_entries/slack-subscription-handler.ts'
      ),
      description: 'Slack subscription Handler',
      environment: {
        CONFIG_PARAM: configParam.parameterName,
      },
    });

    // TODO: allow topic to invoke slack handler
    // slackHandler.addPermission(`AllowInvoke:${Names}`);

    configParam.grantRead(slackHandler);
    configParam.grantWrite(slackHandler);

    return {
      subscriberScope: slackHandler,
      subscriberId: topic.node.id,
      endpoint: slackHandler.functionArn,
      protocol: SubscriptionProtocol.LAMBDA,
      filterPolicy: this.props.filterPolicy,
    };
  }

  private _resolveScopeFromTopic(topic: ITopic): Construct {
    return topic.node.scope as Construct;
  }
}
