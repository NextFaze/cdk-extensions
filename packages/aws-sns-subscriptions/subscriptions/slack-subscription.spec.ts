import { Topic } from '@aws-cdk/aws-sns';
import { Stack } from '@aws-cdk/core';
import { SlackSubscription } from './slack-subscription';
import { SynthUtils } from '@aws-cdk/assert';

let topic: Topic;
let stack: Stack;
beforeEach(() => {
  stack = new Stack();
  topic = new Topic(stack, 'MyTopic');
});

test('configures slack handler and parameter resources', () => {
  topic.addSubscription(
    new SlackSubscription({
      channelName: 'my-channel',
    })
  );

  expect(SynthUtils.toCloudFormation(stack)).toMatchSnapshot();
});
