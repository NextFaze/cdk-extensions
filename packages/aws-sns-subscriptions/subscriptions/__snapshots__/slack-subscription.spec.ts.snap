// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`configures slack handler and parameter resources 1`] = `
Object {
  "Parameters": Object {
    "AssetParametersf75e68ccacb653aa13b40662de6b5088cc3d22a3e4a1d815a4657f1cd26c8aadArtifactHash84E3BFB6": Object {
      "Description": "Artifact hash for asset \\"f75e68ccacb653aa13b40662de6b5088cc3d22a3e4a1d815a4657f1cd26c8aad\\"",
      "Type": "String",
    },
    "AssetParametersf75e68ccacb653aa13b40662de6b5088cc3d22a3e4a1d815a4657f1cd26c8aadS3BucketEC32D104": Object {
      "Description": "S3 bucket for asset \\"f75e68ccacb653aa13b40662de6b5088cc3d22a3e4a1d815a4657f1cd26c8aad\\"",
      "Type": "String",
    },
    "AssetParametersf75e68ccacb653aa13b40662de6b5088cc3d22a3e4a1d815a4657f1cd26c8aadS3VersionKeyB29AB568": Object {
      "Description": "S3 key for asset version \\"f75e68ccacb653aa13b40662de6b5088cc3d22a3e4a1d815a4657f1cd26c8aad\\"",
      "Type": "String",
    },
  },
  "Resources": Object {
    "Handler886CB40B": Object {
      "DependsOn": Array [
        "HandlerServiceRoleDefaultPolicyCBD0CC91",
        "HandlerServiceRoleFCDC14AE",
      ],
      "Properties": Object {
        "Code": Object {
          "S3Bucket": Object {
            "Ref": "AssetParametersf75e68ccacb653aa13b40662de6b5088cc3d22a3e4a1d815a4657f1cd26c8aadS3BucketEC32D104",
          },
          "S3Key": Object {
            "Fn::Join": Array [
              "",
              Array [
                Object {
                  "Fn::Select": Array [
                    0,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParametersf75e68ccacb653aa13b40662de6b5088cc3d22a3e4a1d815a4657f1cd26c8aadS3VersionKeyB29AB568",
                        },
                      ],
                    },
                  ],
                },
                Object {
                  "Fn::Select": Array [
                    1,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParametersf75e68ccacb653aa13b40662de6b5088cc3d22a3e4a1d815a4657f1cd26c8aadS3VersionKeyB29AB568",
                        },
                      ],
                    },
                  ],
                },
              ],
            ],
          },
        },
        "Description": "Slack subscription Handler",
        "Environment": Object {
          "Variables": Object {
            "AWS_NODEJS_CONNECTION_REUSE_ENABLED": "1",
            "CONFIG_PARAM": Object {
              "Ref": "MyChannelConfigParameter092BCC61",
            },
          },
        },
        "Handler": "index.handler",
        "Role": Object {
          "Fn::GetAtt": Array [
            "HandlerServiceRoleFCDC14AE",
            "Arn",
          ],
        },
        "Runtime": "nodejs12.x",
        "Timeout": 180,
      },
      "Type": "AWS::Lambda::Function",
    },
    "HandlerAllowInvokeMyTopic0FE351D2": Object {
      "Properties": Object {
        "Action": "lambda:InvokeFunction",
        "FunctionName": Object {
          "Fn::GetAtt": Array [
            "Handler886CB40B",
            "Arn",
          ],
        },
        "Principal": "sns.amazonaws.com",
        "SourceArn": Object {
          "Ref": "MyTopic86869434",
        },
      },
      "Type": "AWS::Lambda::Permission",
    },
    "HandlerMyTopic7B4D3835": Object {
      "Properties": Object {
        "Endpoint": Object {
          "Fn::GetAtt": Array [
            "Handler886CB40B",
            "Arn",
          ],
        },
        "Protocol": "lambda",
        "TopicArn": Object {
          "Ref": "MyTopic86869434",
        },
      },
      "Type": "AWS::SNS::Subscription",
    },
    "HandlerServiceRoleDefaultPolicyCBD0CC91": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": Array [
                "ssm:DescribeParameters",
                "ssm:GetParameters",
                "ssm:GetParameter",
                "ssm:GetParameterHistory",
              ],
              "Effect": "Allow",
              "Resource": Object {
                "Fn::Join": Array [
                  "",
                  Array [
                    "arn:",
                    Object {
                      "Ref": "AWS::Partition",
                    },
                    ":ssm:",
                    Object {
                      "Ref": "AWS::Region",
                    },
                    ":",
                    Object {
                      "Ref": "AWS::AccountId",
                    },
                    ":parameter/",
                    Object {
                      "Ref": "MyChannelConfigParameter092BCC61",
                    },
                  ],
                ],
              },
            },
            Object {
              "Action": "ssm:PutParameter",
              "Effect": "Allow",
              "Resource": Object {
                "Fn::Join": Array [
                  "",
                  Array [
                    "arn:",
                    Object {
                      "Ref": "AWS::Partition",
                    },
                    ":ssm:",
                    Object {
                      "Ref": "AWS::Region",
                    },
                    ":",
                    Object {
                      "Ref": "AWS::AccountId",
                    },
                    ":parameter/",
                    Object {
                      "Ref": "MyChannelConfigParameter092BCC61",
                    },
                  ],
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "HandlerServiceRoleDefaultPolicyCBD0CC91",
        "Roles": Array [
          Object {
            "Ref": "HandlerServiceRoleFCDC14AE",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "HandlerServiceRoleFCDC14AE": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "MyChannelConfigParameter092BCC61": Object {
      "Properties": Object {
        "Description": "Slack configuration parameter",
        "Type": "String",
        "Value": "{\\"channelId\\":\\"\\",\\"channelName\\":\\"my-channel\\",\\"channelType\\":\\"public_channel\\",\\"authToken\\":\\"\\"}",
      },
      "Type": "AWS::SSM::Parameter",
    },
    "MyTopic86869434": Object {
      "Type": "AWS::SNS::Topic",
    },
  },
}
`;
